Copy-construction from vectors with different template parameters fails

Compiling the following code:
```cpp
#include <linear_algebra.hpp>

int main()
{
    auto v = STD_LA::fs_vector<double, 3>{};
    auto u = STD_LA::fs_vector<int, 3>{v};
    auto w = STD_LA::fs_vector<double, 4>{v};
}
```

Fails with the following error:
```
[rbrugo@vim ~/test] $ g++ -std=c++2a linear_algebra.cpp 
In file included from /usr/local/include/linear_algebra.hpp:74,
                 from linear_algebra.cpp:1:
/usr/local/include/linear_algebra/vector.hpp: In instantiation of ‘constexpr std::experimental::math::vector<ET, OT>::vector(const std::experimental::math::vector<ET2, OT2>&) [with ET2 = std::experimental::math::fs_vector_engine<double, 3>; OT2 = std::experimental::math::matrix_operation_traits; ET = std::experimental::math::fs_vector_engine<int, 3>; OT = std::experimental::math::matrix_operation_traits]’:
linear_algebra.cpp:6:41:   required from here
/usr/local/include/linear_algebra/vector.hpp:226:26: error: no matching function for call to ‘std::experimental::math::fs_vector_engine<int, 3>::fs_vector_engine(const engine_type&)’
  226 | :   m_engine(rhs.m_engine)
      |                          ^
In file included from /usr/local/include/linear_algebra.hpp:68,
                 from linear_algebra.cpp:1:
/usr/local/include/linear_algebra/fixed_size_engines.hpp:46:15: note: candidate: ‘template<class U> constexpr std::experimental::math::fs_vector_engine<T, N>::fs_vector_engine(std::initializer_list<_Up>)’
   46 |     constexpr fs_vector_engine(initializer_list<U> list);
      |               ^~~~~~~~~~~~~~~~
/usr/local/include/linear_algebra/fixed_size_engines.hpp:46:15: note:   template argument deduction/substitution failed:
In file included from /usr/local/include/linear_algebra.hpp:74,
                 from linear_algebra.cpp:1:
/usr/local/include/linear_algebra/vector.hpp:226:26: note:   ‘std::experimental::math::fs_vector_engine<double, 3>’ is not derived from ‘std::initializer_list<_Tp>’
  226 | :   m_engine(rhs.m_engine)
      |                          ^
In file included from /usr/local/include/linear_algebra.hpp:68,
                 from linear_algebra.cpp:1:
/usr/local/include/linear_algebra/fixed_size_engines.hpp:44:15: note: candidate: ‘constexpr std::experimental::math::fs_vector_engine<T, N>::fs_vector_engine(const std::experimental::math::fs_vector_engine<T, N>&) [with T = int; long unsigned int N = 3]’
   44 |     constexpr fs_vector_engine(fs_vector_engine const&) = default;
      |               ^~~~~~~~~~~~~~~~
/usr/local/include/linear_algebra/fixed_size_engines.hpp:44:32: note:   no known conversion for argument 1 from ‘const engine_type’ {aka ‘const std::experimental::math::fs_vector_engine<double, 3>’} to ‘const std::experimental::math::fs_vector_engine<int, 3>&’
   44 |     constexpr fs_vector_engine(fs_vector_engine const&) = default;
      |                                ^~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/linear_algebra/fixed_size_engines.hpp:43:15: note: candidate: ‘constexpr std::experimental::math::fs_vector_engine<T, N>::fs_vector_engine(std::experimental::math::fs_vector_engine<T, N>&&) [with T = int; long unsigned int N = 3]’
   43 |     constexpr fs_vector_engine(fs_vector_engine&&) noexcept = default;
      |               ^~~~~~~~~~~~~~~~
/usr/local/include/linear_algebra/fixed_size_engines.hpp:43:32: note:   no known conversion for argument 1 from ‘const engine_type’ {aka ‘const std::experimental::math::fs_vector_engine<double, 3>’} to ‘std::experimental::math::fs_vector_engine<int, 3>&&’
   43 |     constexpr fs_vector_engine(fs_vector_engine&&) noexcept = default;
      |                                ^~~~~~~~~~~~~~~~~~
/usr/local/include/linear_algebra/fixed_size_engines.hpp:79:1: note: candidate: ‘constexpr std::experimental::math::fs_vector_engine<T, N>::fs_vector_engine() [with T = int; long unsigned int N = 3]’
   79 | fs_vector_engine<T,N>::fs_vector_engine()
      | ^~~~~~~~~~~~~~~~~~~~~
/usr/local/include/linear_algebra/fixed_size_engines.hpp:79:1: note:   candidate expects 0 arguments, 1 provided
In file included from /usr/local/include/linear_algebra.hpp:74,
                 from linear_algebra.cpp:1:
/usr/local/include/linear_algebra/vector.hpp: In instantiation of ‘constexpr std::experimental::math::vector<ET, OT>::vector(const std::experimental::math::vector<ET2, OT2>&) [with ET2 = std::experimental::math::fs_vector_engine<double, 3>; OT2 = std::experimental::math::matrix_operation_traits; ET = std::experimental::math::fs_vector_engine<double, 4>; OT = std::experimental::math::matrix_operation_traits]’:
linear_algebra.cpp:7:44:   required from here
/usr/local/include/linear_algebra/vector.hpp:226:26: error: no matching function for call to ‘std::experimental::math::fs_vector_engine<double, 4>::fs_vector_engine(const engine_type&)’
  226 | :   m_engine(rhs.m_engine)
      |                          ^
In file included from /usr/local/include/linear_algebra.hpp:68,
                 from linear_algebra.cpp:1:
/usr/local/include/linear_algebra/fixed_size_engines.hpp:46:15: note: candidate: ‘template<class U> constexpr std::experimental::math::fs_vector_engine<T, N>::fs_vector_engine(std::initializer_list<_Up>)’
   46 |     constexpr fs_vector_engine(initializer_list<U> list);
      |               ^~~~~~~~~~~~~~~~
/usr/local/include/linear_algebra/fixed_size_engines.hpp:46:15: note:   template argument deduction/substitution failed:
In file included from /usr/local/include/linear_algebra.hpp:74,
                 from linear_algebra.cpp:1:
/usr/local/include/linear_algebra/vector.hpp:226:26: note:   ‘std::experimental::math::fs_vector_engine<double, 3>’ is not derived from ‘std::initializer_list<_Tp>’
  226 | :   m_engine(rhs.m_engine)
      |                          ^
In file included from /usr/local/include/linear_algebra.hpp:68,
                 from linear_algebra.cpp:1:
/usr/local/include/linear_algebra/fixed_size_engines.hpp:44:15: note: candidate: ‘constexpr std::experimental::math::fs_vector_engine<T, N>::fs_vector_engine(const std::experimental::math::fs_vector_engine<T, N>&) [with T = double; long unsigned int N = 4]’
   44 |     constexpr fs_vector_engine(fs_vector_engine const&) = default;
      |               ^~~~~~~~~~~~~~~~
/usr/local/include/linear_algebra/fixed_size_engines.hpp:44:32: note:   no known conversion for argument 1 from ‘const engine_type’ {aka ‘const std::experimental::math::fs_vector_engine<double, 3>’} to ‘const std::experimental::math::fs_vector_engine<double, 4>&’
   44 |     constexpr fs_vector_engine(fs_vector_engine const&) = default;
      |                                ^~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/linear_algebra/fixed_size_engines.hpp:43:15: note: candidate: ‘constexpr std::experimental::math::fs_vector_engine<T, N>::fs_vector_engine(std::experimental::math::fs_vector_engine<T, N>&&) [with T = double; long unsigned int N = 4]’
   43 |     constexpr fs_vector_engine(fs_vector_engine&&) noexcept = default;
      |               ^~~~~~~~~~~~~~~~
/usr/local/include/linear_algebra/fixed_size_engines.hpp:43:32: note:   no known conversion for argument 1 from ‘const engine_type’ {aka ‘const std::experimental::math::fs_vector_engine<double, 3>’} to ‘std::experimental::math::fs_vector_engine<double, 4>&&’
   43 |     constexpr fs_vector_engine(fs_vector_engine&&) noexcept = default;
      |                                ^~~~~~~~~~~~~~~~~~
/usr/local/include/linear_algebra/fixed_size_engines.hpp:79:1: note: candidate: ‘constexpr std::experimental::math::fs_vector_engine<T, N>::fs_vector_engine() [with T = double; long unsigned int N = 4]’
   79 | fs_vector_engine<T,N>::fs_vector_engine()
      | ^~~~~~~~~~~~~~~~~~~~~
/usr/local/include/linear_algebra/fixed_size_engines.hpp:79:1: note:   candidate expects 0 arguments, 1 provided
```

--- a/linear_algebra/code/include/linear_algebra/fixed_size_engines.hpp
+++ b/linear_algebra/code/include/linear_algebra/fixed_size_engines.hpp
@@ -42,6 +42,9 @@ class fs_vector_engine
     constexpr fs_vector_engine();
     constexpr fs_vector_engine(fs_vector_engine&&) noexcept = default;
     constexpr fs_vector_engine(fs_vector_engine const&) = default;
+    template<class T2, size_t N2>
+    constexpr fs_vector_engine(fs_vector_engine<T2, N2> const& src);
+
     template<class U>
     constexpr fs_vector_engine(initializer_list<U> list);
 
@@ -81,6 +84,27 @@ fs_vector_engine<T,N>::fs_vector_engine()
     for (auto& elem : ma_elems)
     {
         elem = static_cast<T>(0);
+    }
+}
+
+template<class T, size_t N>
+template<class T2, size_t N2> constexpr
+fs_vector_engine<T,N>::fs_vector_engine(fs_vector_engine<T2, N2> const& src)
+: ma_elems()
+{
+    size_type   count = min<size_type>(N, N2);
+
+    for (size_type i = 0;  i < count;  ++i)
+    {
+        ma_elems[i] = static_cast<T>(src(i));
+    }
+
+    if (count < N)
+    {
+        for (size_type i = count;  i < N;  ++i)
+        {
+            ma_elems[i] = static_cast<T>(0);
+        }
     }
 }
 

